### ISSUE TITLE
Language modeling loss by passing 'labels' while using ONNX inference session

### ISSUE THREAD

**Issue by SachJbp** [microsoft/onnxruntime#8675]
**Is your feature request related to a problem? Please describe.**
I want to calculate the perplexity score of GPT-2 language model on a text for which I need to have access to language modeling loss on a given text. When using GPT2 we can simply pass on the 'labels' parameter to get the loss as follows:

```
import torch
from transformers import GPT2Tokenizer, GPT2LMHeadModel

tokenizer = GPT2Tokenizer.from_pretrained('gpt2')
model = GPT2LMHeadModel.from_pretrained('gpt2', return_dict=True)

inputs = tokenizer("Hello, my dog is cute", return_tensors="pt")
outputs = model(**inputs, labels=inputs["input_ids"])
loss = outputs.loss
```

But, not able to find out how to get the same loss in an ONNX inference session. I am using the below code which only returns the 'last_hidden_state':

import onnxruntime as ort

from transformers import GPT2TokenizerFast
#tokenizer = GPT2TokenizerFast.from_pretrained("gpt2")

ort_session = ort.InferenceSession("onnx/gpt2/model.onnx")

inputs = tokenizer("Using BERT in ONNX!", return_tensors="np")
outputs = ort_session.run(["last_hidden_state"], dict(inputs))

**Describe the solution you'd like**
Would like to have access to language modeling loss by passing labels as input_ids.

**Describe alternatives you've considered**
Calculating the language modeling loss externally but having errors in it.

**Comment by ytaous** [comment]
Hi, have you checked out the example? https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/python/tools/transformers/notebooks/Inference_GPT2_with_OnnxRuntime_on_CPU.ipynb
per notebook example, looks like outputs[0] is what you are looking for, and you can compare it with torch's loss

**Comment by SachJbp** [comment]
Hi, 

I did see that link but that is just comparing the loss incurred due to quantization in ONNX when compared to torch model inference.

I want the per-token (average) or total log-likelihood for all tokens of a given text to be returned as the loss, in other way a measure of how predictable a sentence is for the language model. Eg, if the sentence is "This looks good", it would be **log(P(This)*P(looks|This)*P(good|This Looks))** . It is also called the "Grammar Score" for traditional language models. https://huggingface.co/transformers/perplexity.html describes it as well and its usage for calculating perplexity. They return a per token (average) log-likelihood.

**Comment by tianleiwu** [comment]
@SachJbp, 
I think it is because the exported onnx model does not have labels in input and loss in the onnx output. You can modify the export script to achieve that. 

The places to add labels to input:
https://github.com/microsoft/onnxruntime/blob/064a385b59598419c287a4b897afca376227d3bd/onnxruntime/python/tools/transformers/gpt2_helper.py#L84-L88

You will need update the export_onnx function to have labels in dummy inputs and loss in output:
https://github.com/microsoft/onnxruntime/blob/064a385b59598419c287a4b897afca376227d3bd/onnxruntime/python/tools/transformers/gpt2_helper.py#L265

**Comment by SachJbp** [comment]
Thanks a lot for the assistance. I was able to make the above changes to make it work and get the "loss" on Colab. But, when trying to build an image with same code, I get this error "cannot import name 'get_all_providers' from 'onnxruntime.capi._pybind_state' " when trying to import 'from onnxruntime.transformers.gpt2_helper import Gpt2Helper, MyGPT2LMHeadMode'. Any suggestions ?

**Comment by TDehaene** [comment]
@SachJbp : could you please post your code snippet to calculate the perplexity of onnx models using onnxruntime? I'm stuck in the same situation.

**Comment by OriAlpha** [comment]
I am also facing same issue, is there any updates regarding this issue??

**Comment by SachJbp** [comment]
I was able to make a change locally in GPTLMHeadModel's forward() method to send `labels = input_ids` which returned `loss` as the first output and it helped solve my problem.

**Comment by OriAlpha** [comment]
Is possible to share code changes?? @SachJbp

**Comment by OriAlpha** [comment]
Will try to compute, thanks
